modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java

```diff
@@ -1735,6 +1719,24 @@ private void executeApplyingTasks(final List<LogEntryAndClosure> tasks) {
                     task.reset();
                     continue;
                 }
+
+                // To prevent safe timestamp values from becoming stale, we must assign them under a valid leader lock.
+                if (task.done instanceof SafeTimeAwareCommandClosure) {
+                    SafeTimeAwareCommandClosure clo = (SafeTimeAwareCommandClosure) task.done;
+                    WriteCommand command = clo.command();
+                    HybridTimestamp timestamp = command.initiatorTime();
+
+                    if (timestamp != null) {
+                        if (safeTs == null) {
+                            safeTs = clock.update(timestamp);
+                        } else if (timestamp.compareTo(safeTs) > 0) {
+                            safeTs = clock.update(timestamp);
+                        }
+
+                        clo.safeTimestamp(safeTs);
+                    }
+                }
```

Comment:
Would it make sense to extract this to a method of its own to minimize intrusion in the original JRaft code?